ledger/src/hardened_unpack.rs

```diff
@@ -0,0 +1,406 @@
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["snapshots", "0", "aa"],
+            tar::EntryType::Regular
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["aaaa"],
+            tar::EntryType::Regular
+        ));
+    }
+
+    #[test]
+    fn test_archive_is_valid_archive_entry() {
+        assert!(is_valid_genesis_archive_entry(
+            &["genesis.bin"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb"],
+            tar::EntryType::Directory
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb", "foo"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb", "foo", "bar"],
+            tar::EntryType::Regular
+        ));
+
+        assert!(!is_valid_genesis_archive_entry(
+            &["aaaa"],
+            tar::EntryType::Regular
+        ));
+    }
+
+    fn with_finalize_and_unpack<C>(archive: tar::Builder<Vec<u8>>, checker: C) -> Result<()>
+    where
+        C: Fn(&mut Archive<BufReader<&[u8]>>, &Path) -> Result<()>,
+    {
+        let data = archive.into_inner().unwrap();
+        let reader = BufReader::new(&data[..]);
+        let mut archive: Archive<std::io::BufReader<&[u8]>> = Archive::new(reader);
+        let temp_dir = tempfile::TempDir::new().unwrap();
+
+        checker(&mut archive, &temp_dir.into_path())
+    }
+
+    fn finalize_and_unpack_snapshot(archive: tar::Builder<Vec<u8>>) -> Result<()> {
+        with_finalize_and_unpack(archive, |a, b| unpack_snapshot(a, b))
+    }
+
+    fn finalize_and_unpack_genesis(archive: tar::Builder<Vec<u8>>) -> Result<()> {
+        with_finalize_and_unpack(archive, |a, b| unpack_genesis(a, b))
+    }
+
+    #[test]
+    fn test_archive_unpack_snapshot_ok() {
+        let mut header = Header::new_gnu();
+        header.set_path("version").unwrap();
+        header.set_size(4);
+        header.set_cksum();
+
+        let data: &[u8] = &[1, 2, 3, 4];
+
+        let mut archive = Builder::new(Vec::new());
+        archive.append(&header, data).unwrap();
+
+        let result = finalize_and_unpack_snapshot(archive);
+        assert_matches!(result, Ok(()));
+    }
+
+    #[test]
+    fn test_archive_unpack_genesis_ok() {
+        let mut header = Header::new_gnu();
+        header.set_path("genesis.bin").unwrap();
+        header.set_size(4);
+        header.set_cksum();
+
+        let data: &[u8] = &[1, 2, 3, 4];
+
+        let mut archive = Builder::new(Vec::new());
+        archive.append(&header, data).unwrap();
+
+        let result = finalize_and_unpack_genesis(archive);
+        assert_matches!(result, Ok(()));
+    }
+
+    #[test]
+    fn test_archive_unpack_snapshot_invalid_path() {
+        let mut header = Header::new_gnu();
+        // bypass the sanitization of the .set_path()
+        for (p, c) in header
```

Comment:
Maybe add a check for an archive that contains `/etc/passwd`?

Reply:
It should check that all files are relative paths and also contain no `..` paths. I'm not sure if you can have absolute path in the archive.

Reply:
This check is kind of an extra protection, so this doesn't need to be exhaustive. Maybe I should comemnt so.. `..` are excluded by [us](https://github.com/solana-labs/solana/pull/8959/files#diff-17fc305f5509f475e804ba989bb306caR87) and also by the `tar` crate. I did exclude `..` so as to easing our reasoning for the matcher `fn`s. Maybe, I should add a explaining comment.

Also, absolute paths are neutralized by the `tar` crate, I've checked that. However, maybe adding extra `/etc/passwd` case don't hut much. So, I'll just do it.

Reply:
I believe I've done good enough clarification for this in the codebase: b2df391fe33db1157de9771de627f2d22a8cbe19

Any more questions? :)